api.planx.uk/shared/middleware/validate.ts

```diff
@@ -0,0 +1,39 @@
+import { Request, RequestHandler, Response, NextFunction } from "express";
+import { AnyZodObject, ZodSchema, z } from "zod";
+
+/**
+ * Middleware to validate incoming requests to the API
+ * Takes a ZodSchema and returns a validated, and typed, Request object
+ */
+export const validate =
+  (schema: AnyZodObject) =>
+  async (
+    req: Request<z.infer<typeof schema>>,
+    res: Response,
+    next: NextFunction,
+  ) => {
+    try {
+      schema.parse({
+        params: req.params,
+        body: req.body,
+        query: req.query,
+      });
+      return next();
+    } catch (error) {
+      return res.status(400).json(error);
```

Comment:
We could do a lot more with this error, like format it in a more meaningful way with https://github.com/causaly/zod-validation-error but a raw Zod error is meaningful enough for our current API users.